### PR TITLE
Per-attribute transition properties on MGLStyleLayer

### LINKED ISSUE #8010: Per-attribute transition properties on MGLStyleLayer
MGLStyleLayer’s concrete subclasses should allow the developer to set transitions that affect individual layout or paint attributes.

We could pair every layout or paint attribute with a `transitionFor…` property. Or we could fight property proliferation by adding `-transitionForKey:` and `-setTransition:forKey:` methods to MGLStyleLayer itself (along the lines of `-[CALayer addAnimation:forKey:]`). (The downside would be pushing validation of the key until runtime.)

Either way, we should add an MGLTransition class or struct to encapsulate the details of a transition. We’ll want to refine the MGLStyle methods added in #7711 to make use of MGLTransition instead of working with durations and delays directly.

/ref #7863
/cc @incanus @boundsj @jfirebaugh @tobrun

### PR DESCRIPTION
**Pull request by fabian-guerra**
Fixes #8010

### CODE REVIEW

File: platform/darwin/src/MGLTypes.h [hunk 1]
```diff
@@ -85,6 +85,21 @@ typedef NS_OPTIONS(NSUInteger, MGLMapDebugMaskOptions) {
 #endif
 };
 
+/**
+ A structure containing information about a transition values.
```
Comment (incanus) [inline platform/darwin/src/MGLTypes.h]: Remove `values`.

File: platform/darwin/src/MGLTypes.h [hunk 2]
```diff
@@ -85,6 +85,21 @@ typedef NS_OPTIONS(NSUInteger, MGLMapDebugMaskOptions) {
...
+ */
+typedef struct MGLTransition {
+    /**
+     The duration in seconds to animate any changes to the style URL or to layout and paint attributes.
+     */
+    NSTimeInterval duration;
+    
+    /**
+     The delay in seconds to before applying any changes to the style URL or to layout and paint attributes.
```
Comment (incanus) [inline platform/darwin/src/MGLTypes.h]: Remove `to` in `to before`.

File: platform/darwin/src/MGLSymbolStyleLayer.mm [hunk 3]
```diff
@@ -887,13 +889,45 @@ - (void)setTextTransform:(MGLStyleValue<NSValue *> *)textTransform {
 
 #pragma mark - Accessing the Paint Attributes
 
+- (void)setTransition:(MGLTransition)transition forKey:(NSString *)key
+{
+    NSString *camelCaseKey;
+    if ([key length] > 1) {
+        camelCaseKey = [NSString stringWithFormat:@"%@%@",[[key substringToIndex:1] uppercaseString],[key substringFromIndex:1]];
```
Comment (incanus) [inline platform/darwin/src/MGLSymbolStyleLayer.mm]: Code style: spaces after `,`.

File: platform/darwin/src/MGLStyleLayer.mm.ejs [hunk 4]
```diff
@@ -14,6 +14,8 @@
 #import "MGLStyleValue_Private.h"
 #import "MGL<%- camelize(type) %>StyleLayer.h"
 
+#import "NSDate+MGLAdditions.h"
+
```
Comment (incanus) [inline platform/darwin/src/MGLStyleLayer.mm.ejs]: Put this further up with the others, next to `#import "NSPredicate+MGLAdditions.h"`, a similar category use.

File: platform/darwin/src/MGLStyle.h [hunk 5]
```diff
@@ -30,6 +30,7 @@ NS_ASSUME_NONNULL_BEGIN
  */
 static MGL_EXPORT const NSInteger MGLStyleDefaultVersion = 9;
 
+
```
Comment (incanus) [inline platform/darwin/src/MGLStyle.h]: Extraneous whitespace.

File: platform/darwin/src/MGLStyle.h [hunk 6]
```diff
@@ -194,6 +195,11 @@ MGL_EXPORT
 @property (nonatomic, strong) NS_SET_OF(__kindof MGLSource *) *sources;
 
 /**
+ Transition values.
```
Comment (incanus) [inline platform/darwin/src/MGLStyle.h]: Should describe it a bit more here, maybe: 

> Values describing animated transitions to styling changes, either to the style URL or to individual properties.

File: platform/darwin/src/MGLStyle.mm [hunk 7]
```diff
@@ -572,6 +572,20 @@ - (MGLImage *)imageForName:(NSString *)name
 
 #pragma mark Style transitions
 
+- (void)setTransition:(MGLTransition)transition
+{
+    [self setTransitionDuration:transition.duration];
+    [self setTransitionDelay:transition.delay];
```
Comment (incanus) [inline platform/darwin/src/MGLStyle.mm]: We should probably get rid of the separate internal `-setTransitionDuration:` and `-setTransitionDelay:` since externally we only set full transition structures containing both values.

File: platform/darwin/src/MGLStyleLayer.mm.ejs [hunk 8]
```diff
@@ -229,6 +231,31 @@ namespace mbgl {
...
+        NSInvocation *setPropertyInvocation = [NSInvocation
+                                                invocationWithMethodSignature:
+                                                [MGL<%- camelize(type) %>StyleLayer instanceMethodSignatureForSelector:setPropertyTransitionSelector]];
+
+        [setPropertyInvocation setSelector:setPropertyTransitionSelector];
+        [setPropertyInvocation setTarget:self];
+
+        [setPropertyInvocation setArgument:&transition atIndex:2];
+
+        [setPropertyInvocation performSelector:@selector(invoke)];
```
Comment (incanus) [inline platform/darwin/src/MGLStyleLayer.mm.ejs]: Couldn't this be simply `-[NSObject performSelector:withObject:]` since it's only one argument, passing the `transition` wrapped in `NSValue`?

File: platform/darwin/src/MGLStyleLayer.mm.ejs [hunk 9]
```diff
@@ -229,6 +231,31 @@ namespace mbgl {
...
+    NSString *camelCaseKey;
+    if ([key length] > 1) {
+        camelCaseKey = [NSString stringWithFormat:@"%@%@",[[key substringToIndex:1] uppercaseString],[key substringFromIndex:1]];
+    } else {
+        camelCaseKey = [key uppercaseString];
+    }
+    NSString *setPropertyTransitionString = [NSString stringWithFormat:@"mbx_set%@Transition:", camelCaseKey];
+    SEL setPropertyTransitionSelector = NSSelectorFromString(setPropertyTransitionString);
+
+    if ([self respondsToSelector:setPropertyTransitionSelector]) {
```
Comment (boundsj) [inline platform/darwin/src/MGLStyleLayer.mm.ejs]: What is the main motivation to introduce dynamic method invocation for transitions? 

Most of the complexity here is about parsing the string and forming the private method name to match the actual implementation and that introduces new types of edge cases that we did not previously need to handle. For example, what if the client app passes in a nonsensical string at runtime? So far, I think this implementation would fail silently. Should it log a warning, raise an exception, etc.? Do we expect to offer a string parameter based getter for each transition, too?

Since the style layer implementation is generated and in the spirit of consistency with the existing API and its per property getters and setters, I suggest that we consider making the `mbx_` methods in this implementation public and avoid the dynamic invocation altogether.

File: platform/darwin/src/MGLSymbolStyleLayer.h [hunk 10]
```diff
@@ -1204,6 +1206,15 @@ MGL_EXPORT
  * `MGLCameraStyleFunction` with an interpolation mode of:
    * `MGLInterpolationModeExponential`
    * `MGLInterpolationModeInterval`
+ * `MGLSourceStyleFunction` with an interpolation mode of:
```
Comment (boundsj) [inline platform/darwin/src/MGLSymbolStyleLayer.h]: This is a byproduct of the style spec / gl native mismatch. All of these comments should be manually removed in this PR (or work against a style spec gitsha that does not cause this to happen to avoid the inconvenience)

**Comment by boundsj** [comment]
> @incanus Should this PR be based against the 3.5.0 release branch since it builds upon data-driven styling?

The `3.5.0` is actually currently `master`. We expect to make (and share with Android) a 3.5.0 release branch later today or tomorrow. We can point this PR to that branch once it exists.

File: platform/darwin/src/MGLSymbolStyleLayer.h (hunk 10 above)
Comment (boundsj) [inline platform/darwin/src/MGLSymbolStyleLayer.h]: Also, it is likely that https://github.com/mapbox/mapbox-gl-native/pull/7939 will land before this PR and, I think, this problem will go away then once this branch is rebased on top of that change set.

Comment (boundsj) [inline platform/darwin/src/MGLSymbolStyleLayer.h]: Yes, if you rebase this PR on top of master as of now, f901e77 will fix this problem 👍

File: platform/darwin/src/MGLStyleLayer.mm.ejs (hunk 9 above)
Comment (boundsj) [inline platform/darwin/src/MGLStyleLayer.mm.ejs]: I was pointed to the original post in https://github.com/mapbox/mapbox-gl-native/issues/8010#issue-206634577 so this makes sense now. Thanks!

I still vote for explicit getters and setters. However, if we do go with the CALayer influenced API, it would be nice to try a refactor so that some or all of this string parsing and method invocation logic goes into the base MGLStyleLayer class.

File: platform/darwin/src/MGLStyleLayer.mm.ejs (hunk 8 above)
Comment (fabian-guerra) [inline platform/darwin/src/MGLStyleLayer.mm.ejs]: @incanus using an `NSValue` I will have to wrap it then unwrap in `set***Transition:`, also I will have to add pragma marks to avoid the "PerformSelect may cause a leak because...".

I think this approach is more straightforward, what are your thoughts?

File: platform/darwin/src/MGLStyleLayer.mm [hunk 11]
```diff
@@ -65,6 +65,56 @@ - (float)minimumZoomLevel
...
+    NSString *setPropertyTransitionString = [NSString stringWithFormat:@"mbx_%@%@Transition%@", type, camelCaseKey, parameters];
+    SEL propertyTransitionSelector = NSSelectorFromString(setPropertyTransitionString);
+    
+    return propertyTransitionSelector;
+}
+
+- (NSArray *)transitionKeys
+{
+    // This is overridden by subclasses
+    return nil;
```
Comment (1ec5) [inline platform/darwin/src/MGLStyleLayer.mm]: Raise an MGLAbstractClassException. Search for that string for an example to follow.

File: platform/darwin/src/MGLBackgroundStyleLayer.mm [hunk 12]
```diff
@@ -84,13 +85,33 @@ - (void)removeFromMapView:(MGLMapView *)mapView
...
 #pragma mark - Accessing the Paint Attributes
 
+- (NSArray *)transitionKeys
+{
+    NSMutableArray *transitionKeys = [NSMutableArray array];
+    [transitionKeys addObject:@"backgroundColor"];
+    [transitionKeys addObject:@"backgroundOpacity"];
+    [transitionKeys addObject:@"backgroundPattern"];
+
+    return transitionKeys;
```
Comment (1ec5) [inline platform/darwin/src/MGLBackgroundStyleLayer.mm]: It would be more efficient to return an array literal.

File: platform/darwin/src/MGLBackgroundStyleLayer.mm [hunk 13]
```diff
@@ -84,13 +85,33 @@ - (void)removeFromMapView:(MGLMapView *)mapView
 
 #pragma mark - Accessing the Paint Attributes
 
+- (NSArray *)transitionKeys
```
Comment (1ec5) [inline platform/darwin/src/MGLBackgroundStyleLayer.mm]: Since this method doesn’t have any dependencies on other instance methods, you can make this a class method. But I’m unclear on this method’s purpose. It appears to be unused.

**Comment by 1ec5** [comment]
Since this PR is scheduled for iOS SDK v3.5.0, please rebase and retarget it at the release-ios-v3.5.0-android-v5.0.0 branch. Thanks!

File: platform/darwin/src/MGLBackgroundStyleLayer.mm (hunk 13 above)
Comment (fabian-guerra) [inline platform/darwin/src/MGLBackgroundStyleLayer.mm]: This method returns the keys which you can add a transition

File: platform/darwin/src/MGLStyleLayer.mm.ejs [hunk 14]
```diff
@@ -229,6 +230,19 @@ namespace mbgl {
...
+<% for (const property of paintProperties) { -%>
+    [transitionKeys addObject:@"<%- camelizeWithLeadingLowercase(property.name) %>"];
+<% if (property.original) { -%>
+    [transitionKeys addObject:@"<%- camelizeWithLeadingLowercase(originalPropertyName(property)) %>"];
+<% } -%>
+<% } -%>
+
+    return transitionKeys;
+}
+
```
Comment (incanus) [inline platform/darwin/src/MGLStyleLayer.mm.ejs]: Yeah, per https://github.com/mapbox/mapbox-gl-native/pull/8225/files#r104082076, you could do like this: 

```objc
+ (NSArray *)transitionKeys
{    
    return @[
<% for ...
       @"<%- camelizeWithLeadingLowercase(property.name) %>",
<% } -%>
    ];
}
```

Comment (fabian-guerra) [inline platform/darwin/src/MGLStyleLayer.mm.ejs]: @incanus I was using master but retarget to rebase release-ios-v3.5.0-android-v5.0.0

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/Mapbox.java [hunk 15]
```diff
@@ -43,7 +43,7 @@ public static synchronized Mapbox getInstance(@NonNull Context context, @NonNull
     return INSTANCE;
   }
 
-  private Mapbox(@NonNull Context context, @NonNull String accessToken) {
+  Mapbox(@NonNull Context context, @NonNull String accessToken) {
```
Comment (boundsj) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/Mapbox.java]: This PR seems to have picked up additional, unrelated commits along the way when it was rebased like this one from https://github.com/mapbox/mapbox-gl-native/pull/8228

**Comment by incanus** [comment]
Fixed https://github.com/mapbox/mapbox-gl-native/pull/8225#pullrequestreview-25405304 manually with a rebase dropping these oddball non-@fabian-guerra commits and force-pushed to fix 👍

File: platform/darwin/src/MGLStyleLayer.mm.ejs (hunk 14 above)
Comment (1ec5) [inline platform/darwin/src/MGLStyleLayer.mm.ejs]: In the future, do `git rebase -i newBranchName` and remove things you don't want as part of the PR. Then force-push and finally retarget the PR.

File: platform/darwin/src/MGLStyleLayer.h [hunk 16]
```diff
@@ -69,6 +69,17 @@ MGL_EXPORT
...
+#pragma mark Layer Transitions
+
+- (void)setTransition:(MGLTransition)transition forKey:(NSString *)key;
+
+- (MGLTransition)transitionForKey:(NSString *)key;
+
+/**
+ Returns an array of strings that identify transition keys attached to the layer.
+ */
++ (NSArray *)transitionKeys;
```
Comment (1ec5) [inline platform/darwin/src/MGLStyleLayer.h]: Should we expose this method publicly? If so, let's call it `transitionableAttributes` and also make the return type a set of strings instead of an unqualified array.

File: platform/darwin/src/MGLStyleLayer.h [hunk 17]
```diff
@@ -69,6 +69,17 @@ MGL_EXPORT
  */
 @property (nonatomic, assign) float minimumZoomLevel;
 
+#pragma mark Layer Transitions
```
Comment (1ec5) [inline platform/darwin/src/MGLStyleLayer.h]: jazzy uses these marks to produce section headings in the generated documentation. For consistency with Apple's documentation, we label each mark according to a task. Here, we can say "Timing Changes to Attributes".

File: platform/darwin/src/MGLStyleLayer.h [hunk 18]
```diff
@@ -69,6 +69,17 @@ MGL_EXPORT
...
+
+- (void)setTransition:(MGLTransition)transition forKey:(NSString *)key;
```
Comment (1ec5) [inline platform/darwin/src/MGLStyleLayer.h]: Public methods need documentation comments.

Comment (1ec5) [inline platform/darwin/src/MGLStyleLayer.h]: I'm thinking we might want to call it `-setTransition:forAttribute:` to be more specific about the kind of key that's being transitioned.

File: platform/darwin/src/MGLStyleLayer.mm [hunk 19]
```diff
@@ -65,6 +65,58 @@ - (float)minimumZoomLevel
...
+#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
+        NSValue *transitionValue = [self performSelector:getPropertyTransitionSelector];
+        [transitionValue getValue:&transition];
+#pragma clang diagnostic pop
+    }
+    
+    return transition;
+}
+
+- (SEL)mbx_selectorForKey:(NSString *)key type:(NSString *)type hasParams:(BOOL)params
```
Comment (1ec5) [inline platform/darwin/src/MGLStyleLayer.mm]: There's no need to prefix private methods with a class prefix the way we prefix category methods.

Comment (1ec5) [inline platform/darwin/src/MGLStyleLayer.mm]: This method would be a bit simpler if we had two methods: one for getters and one for setters.

Comment (1ec5) [inline platform/darwin/src/MGLStyleLayer.mm]: Since all the valid setter method names form a closed set, we should generate a lookup table mapping attributes to the corresponding setter name, then look up the passed-in attribute in that table to decide what to call. That way, we can raise an exception when the developer passes in an untransitionable attribute or a bogus attribute. Currently, doing that leads to a nasty crash with no useful information.

File: platform/darwin/src/MGLBackgroundStyleLayer.mm (hunk 13 above)
Comment (1ec5) [inline platform/darwin/src/MGLBackgroundStyleLayer.mm]: Ah, looking more closely, this method appears to be used only in tests (but is exposed publicly). As an alternative to this method, could we move the array of keys into the tests themselves?

File: platform/darwin/src/MGLStyleLayer.h (hunk 16 above)
Comment (fabian-guerra) [inline platform/darwin/src/MGLStyleLayer.h]: This Method is intended to explicitly return which layer attributes are transitionable, we can as well add a note on each property, but I think this is a better way to let know devs about what they can change, also this can be used as helper to set the same transition property to all transitionable attributes

**Comment by fabian-guerra** [comment]
After talking to @1ec5 and @boundsj I realized that the trade-off for having a single method to set transition VS a straightforward approach was not enough to justify the added complexity.

Each transitionable style attribute has it's corresponding method.

File: platform/darwin/src/MGLBackgroundStyleLayer.h [hunk 20]
```diff
@@ -69,6 +69,11 @@ MGL_EXPORT
 #endif
 
 /**
+ `backgroundColor` transition attributes.
```
Comment (1ec5) [inline platform/darwin/src/MGLBackgroundStyleLayer.h]: > The transition affecting any changes to this layer’s `backgroundColor` property.
>
> This property corresponds to the `background-color-transition` property in the style JSON file format.

(We can't quite say "in the Mapbox Style Specification", as we do elsewhere, because of mapbox/mapbox-gl-js#4081.)

File: platform/darwin/test/MGLBackgroundStyleLayerTests.mm [hunk 21]
```diff
@@ -59,6 +62,11 @@ - (void)testProperties {
         XCTAssertThrowsSpecificNamed(layer.backgroundColor = functionStyleValue, NSException, NSInvalidArgumentException, @"MGLStyleValue should raise an exception if it is applied to a property that cannot support it");
         functionStyleValue = [MGLStyleValue<MGLColor *> valueWithInterpolationMode:MGLInterpolationModeInterval compositeStops:@{@18: constantStyleValue} attributeName:@"" options:nil];
         XCTAssertThrowsSpecificNamed(layer.backgroundColor = functionStyleValue, NSException, NSInvalidArgumentException, @"MGLStyleValue should raise an exception if it is applied to a property that cannot support it");
+        // Transition property test
+        layer.backgroundColorTransition = transitionTest;
+        MGLTransition backgroundColorTransition = layer.backgroundColorTransition;
```
Comment (1ec5) [inline platform/darwin/test/MGLBackgroundStyleLayerTests.mm]: By setting and getting before asserting, we're verifying that the transition value round-trips. That's good, but it doesn't rule out the possibility that both the getter and setter are buggy. To rule that out, we'd need to assert after setting that the underlying `mbgl::style::BackgroundLayer::getBackgroundColorTransition()` is correct.

File: platform/darwin/src/MGLSymbolStyleLayer.h [hunk 22]
```diff
@@ -1753,6 +1814,7 @@ MGL_EXPORT
  */
 @property (nonatomic, null_resettable) MGLStyleValue<NSValue *> *textTranslationAnchor;
 
+
```
Comment (1ec5) [inline platform/darwin/src/MGLSymbolStyleLayer.h]: Nit: stray line.

File: platform/darwin/src/MGLTypes.h [hunk 23]
```diff
@@ -70,6 +70,21 @@ typedef NS_OPTIONS(NSUInteger, MGLMapDebugMaskOptions) {
...
+     The duration in seconds to animate any changes to the style URL or to layout and paint attributes.
+     */
+    NSTimeInterval duration;
+    
+    /**
+     The delay in seconds before applying any changes to the style URL or to layout and paint attributes.
+     */
+    NSTimeInterval delay;
+} MGLTransition;
+
```
Comment (1ec5) [inline platform/darwin/src/MGLTypes.h]: We should provide a factory method MGLTransitionMake(). We might also want to add transition-related methods to NSValue(MGLAdditions) to make it easier to work with MGLTransition instances when an object is needed.

File: platform/darwin/src/MGLTypes.h [hunk 24]
```diff
@@ -70,6 +70,21 @@ typedef NS_OPTIONS(NSUInteger, MGLMapDebugMaskOptions) {
 #endif
 };
 
+/**
+ A structure containing information about a transition.
+ */
+typedef struct MGLTransition {
+    /**
+     The duration in seconds to animate any changes to the style URL or to layout and paint attributes.
```
Comment (1ec5) [inline platform/darwin/src/MGLTypes.h]: It's a bit of a nonsequitur to talk about style URLs and attributes here. This structure is generic enough that its documentation should be more self-contained. For example:

> The amount of time the animation should take, not including the `delay`.

(It doesn't include the delay, does it?)

File: platform/darwin/src/MGLTypes.h [hunk 25]
```diff
@@ -70,6 +70,21 @@ typedef NS_OPTIONS(NSUInteger, MGLMapDebugMaskOptions) {
...
+     */
+    NSTimeInterval duration;
+    
+    /**
+     The delay in seconds before applying any changes to the style URL or to layout and paint attributes.
```
Comment (1ec5) [inline platform/darwin/src/MGLTypes.h]: > The amount of time in seconds to wait before beginning the animation.

File: platform/darwin/src/MGLBackgroundStyleLayer.h [hunk 26]
```diff
@@ -69,6 +69,13 @@ MGL_EXPORT
 #endif
 
 /**
+ The transition affecting any changes to this layer’s `backgroundColor` property.
+
+ This property corresponds to the background-color-transition property in the style JSON file format.
```
Comment (1ec5) [inline platform/darwin/src/MGLBackgroundStyleLayer.h]: Surround `background-color-transition` in backticks.

File: platform/darwin/test/MGLFillStyleLayerTests.mm [hunk 27]
```diff
@@ -137,6 +142,18 @@ - (void)testProperties {
...
         XCTAssertEqualObjects(layer.fillColor, defaultStyleValue,
                               @"fillColor should return the default value after being unset.");
+        // Transition property test
+        layer.fillColorTransition = transitionTest;
+        MGLTransition options;
+        auto toptions = rawLayer->getFillColorTransition();
+        options.delay = MGLTimeIntervalFromDuration(toptions.delay.value_or(mbgl::Duration::zero()));
+        options.duration = MGLTimeIntervalFromDuration(toptions.duration.value_or(mbgl::Duration::zero()));
+        XCTAssertEqual(options.delay, transitionTest.delay);
+        XCTAssertEqual(options.duration, transitionTest.duration);
```
Comment (1ec5) [inline platform/darwin/test/MGLFillStyleLayerTests.mm]: I would structure this a little differently:

* Set `fillColorTransition` to `transitionTest`.
* Call `getFillColorTransition()`.
* Assert `toptions.delay && *toptions.delay == transitionTest.delay`.

File: platform/darwin/src/MGLTypes.h [hunk 28]
```diff
@@ -70,6 +70,32 @@ typedef NS_OPTIONS(NSUInteger, MGLMapDebugMaskOptions) {
...
+    
+    /**
+     The amount of time in seconds to wait before beginning the animation.
+     */
+    NSTimeInterval delay;
+} MGLTransition;
+
+/**
+ Creates a new `MGLTransition` from the given duration and delay.
+ */
```
Comment (1ec5) [inline platform/darwin/src/MGLTypes.h]: Document each `@param` and the `@return` value.

File: platform/darwin/src/NSValue+MGLAdditions.h [hunk 29]
```diff
@@ -70,6 +71,22 @@ NS_ASSUME_NONNULL_BEGIN
  */
 @property (readonly) MGLOfflinePackProgress MGLOfflinePackProgressValue;
 
+#pragma mark Working with Transition values
```
Comment (1ec5) [inline platform/darwin/src/NSValue+MGLAdditions.h]: Typo: s/values/Values/

File: platform/darwin/src/MGLTypes.h [hunk 30]
```diff
@@ -70,6 +70,39 @@ typedef NS_OPTIONS(NSUInteger, MGLMapDebugMaskOptions) {
...
+ Creates a new `MGLTransition` from the given duration and delay.
+ 
+ @param duration The amount of time the animation should take, not including 
+ the delay.
+ @param delay The amount of time in seconds to wait before beginning the 
+ animation.
+ 
+ @return Returns a `MGLTransition` struct containing the transition attributes.
+ */
+NS_INLINE MGLTransition MGLTransitionMake(NSTimeInterval duration, NSTimeInterval delay) {
```
Comment (1ec5) [inline platform/darwin/src/MGLTypes.h]: I'm not entirely sure, but you may also need to MGL_EXPORT this function. Generally we have to export public, top-level symbols, but this one is inlined. @kkaefer would know for sure.

File: platform/darwin/src/NSValue+MGLAdditions.m [hunk 31]
```diff
@@ -46,4 +46,16 @@ - (MGLOfflinePackProgress)MGLOfflinePackProgressValue {
...
++ (NSValue *)valueWithMGLTransition:(MGLTransition)transition; {
+    return [NSValue value:&transition withObjCType:@encode(MGLTransition)];
+}
+
+- (MGLTransition)MGLTransitionValue {
+    MGLTransition transition;
+    [self getValue:&transition];
+    return transition;
+}
+
```
Comment (incanus) [inline platform/darwin/src/NSValue+MGLAdditions.m]: Why have these utilities anymore? We don't use them.

Comment (incanus) [inline platform/darwin/src/NSValue+MGLAdditions.m]: I've been pointed at https://github.com/mapbox/mapbox-gl-native/pull/8225#discussion_r104857840. 👍🏻

**Comment by incanus** [comment]
🎉
